utils/tests_fetcher.py

```diff
@@ -254,6 +256,122 @@ def diff_contains_doc_examples(repo: Repo, branching_point: str, filename: str)
     return old_content_clean != new_content_clean
 
 
+def get_impacted_files_from_tiny_model_summary(diff_with_last_commit: bool = False) -> List[str]:
+    """
+    Return a list of python modeling files that are impacted by the changes of `tiny_model_summary.json` in between:
+
+    - the current head and the main branch if `diff_with_last_commit=False` (default)
+    - the current head and its parent commit otherwise.
+
+    Returns:
+        `List[str]`: The list of Python modeling files that are impacted by the changes of `tiny_model_summary.json`.
+    """
+    repo = Repo(PATH_TO_REPO)
+
+    folder = Path(repo.working_dir)
+
+    if not diff_with_last_commit:
+        print(f"main is at {repo.refs.main.commit}")
+        print(f"Current head is at {repo.head.commit}")
+
+        commits = repo.merge_base(repo.refs.main, repo.head)
+        for commit in commits:
+            print(f"Branching commit: {commit}")
+    else:
+        print(f"main is at {repo.head.commit}")
+        commits = repo.head.commit.parents
+        for commit in commits:
+            print(f"Parent commit: {commit}")
+
+    if not os.path.isfile(folder / "tests/utils/tiny_model_summary.json"):
+        return []
+
+    files = set()
+    for commit in commits:
+        with checkout_commit(repo, commit):
+            with open(folder / "tests/utils/tiny_model_summary.json", "r", encoding="utf-8") as f:
+                old_content = f.read()
+
+        with open(folder / "tests/utils/tiny_model_summary.json", "r", encoding="utf-8") as f:
+            new_content = f.read()
+
+        # get the content as json object
+        old_content = json.loads(old_content)
+        new_content = json.loads(new_content)
+
+        old_keys = set(old_content.keys())
+        new_keys = set(new_content.keys())
+
+        # get the difference
+        diff_keys = old_keys.symmetric_difference(new_keys)
+        common_keys = old_keys.intersection(new_keys)
+        # if both have the same key, check its content
+        for key in common_keys:
+            if old_content[key] != new_content[key]:
+                diff_keys.add(key)
+
+        # get the model classes
+        impacted_model_classes = []
+        for key in diff_keys:
+            if key in new_keys:
+                impacted_model_classes.extend(new_content[key]["model_classes"])
+
+        # get the module where the model classes are defined. We want to use the main `__init__` file, but it requires
+        # all the framework being installed, which is not ideal for a simple script like test fetcher.
+        # So we create a temporary and modified main `__init__` and access its `_import_structure`.
```

Comment:
why not use inspect on the class? The transformers packages should be installed and 
```python 
>>> import inspect
>>> from transformers import *
>>> inspect.getfile(globals()["CLIPConfig"])
 '/Users/arthurzucker/Work/transformers/src/transformers/models/clip/configuration_clip.py'
```
we can split on `src` and be done with it?

Reply:
Using` inspect` will make this `test_fethcer` require all the dependency being installed, otherwise one will get a dummy object, also the process will become slower due to all the importing.

Reply:
> more resilient less error prone

I agree too.

I don't think the community would run test fetcher. So I am fine if we decide to use `inspect`  (on the CI). 

- On `CircleCI`: the fetch test job might take longer (bigger cache to download, longer installation if cache miss)
- On `Push CI`: the results would look a bit strange/confusing, as `Flax` is not installed in our CI docker image.

Reply:
If I understood correctly we need dependencies anyway no? So what we can do is:
- load the object
- if it's a dummy, log that dummy object and don't use it (will help you with debugging). That means we don't have the deps on the machine
- else, we have the deps, all good

Reply:
currently , test fetcher run in ~ 1m30.

Loading cache of the whole `.[dev]` will take 5 minutes + a bit more for all the importing

Reply:
Fetching the test should be fast, so would just do this job last. I thought we had a different machine running the pipeline test and would say let's forward getting the files to update over there if there are any changes to the tiny model! 
Cache is so slow :(

Reply:
>  forward getting the files to update over there 

If the commit(s) in a PR only modifies the `tests/utils/tiny_model_summary.json`, then `fetch_tests` in `.circleci/config.yml` will give an empty job. So there is no job (like `torch_job` / `tf_job`) to perform `getting the files to update`, and we will miss the desired pipeline tests to run.

Reply:
Thanks for the offline discussion. Yes, your suggestion is a good possibility (other than the fact the file `.circleci/create_circleci_config.py` might get a bit more complex in order to get the actual tests to run dynamically).

As mentioned, I proposed to merge the current version, and I will work in a follow up PR to implement your suggested idea. So at the end, if there is any unexpected undesired issue, we can at least have a working version.